framework/log/FileTarget.php

```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
         for ($i = $this->maxLogFiles; $i >= 0; --$i) {
             // $i == 0 is the original log file
             $rotateFile = $file . ($i === 0 ? '' : '.' . $i);
-            if (is_file($rotateFile)) {
-                // suppress errors because it's possible multiple processes enter into this section
-                if ($i === $this->maxLogFiles) {
-                    @unlink($rotateFile);
+            $rotateTarget = $file . '.' . ($i + 1);
+            if (!is_file($rotateFile)) {
+                if ($this->compressRotatedFiles && is_file("$rotateFile.gz")) {
+                    $rotateFile = "$rotateFile.gz";
+                    $rotateTarget = "$rotateTarget.gz";
                 } else {
-                    if ($this->rotateByCopy) {
-                        @copy($rotateFile, $file . '.' . ($i + 1));
-                        if ($fp = @fopen($rotateFile, 'a')) {
-                            @ftruncate($fp, 0);
-                            @fclose($fp);
-                        }
-                        if ($this->fileMode !== null) {
-                            @chmod($file . '.' . ($i + 1), $this->fileMode);
-                        }
-                    } else {
-                        @rename($rotateFile, $file . '.' . ($i + 1));
-                    }
+                    continue;
                 }
             }
+            if ($i === $this->maxLogFiles) {
+                // suppress errors because it's possible multiple processes enter into this section
+                @unlink($rotateFile);
+            } else {
+                $this->rotateFile($rotateFile, $rotateTarget);
+            }
+        }
+    }
+
+    /**
+     * Rotate a single file
+     * @param string $file the file to rotate.
+     * @param string $target the target file name.
+     */
+    protected function rotateFile($file, $target)
+    {
+        // suppress errors because it's possible multiple processes enter into this section
+        if ($this->rotateByCopy) {
+            @copy($file, $target);
+            if ($fp = @fopen($file, 'a')) {
+                @ftruncate($fp, 0);
+                @fclose($fp);
+            }
+            if ($this->fileMode !== null) {
+                @chmod($target, $this->fileMode);
+            }
+        } else {
+            @rename($file, $target);
+        }
+        if ($this->compressRotatedFiles) {
+            $this->compressFile($target);
+        }
+    }
+
+
+    /**
+     * Compresses a file with gzip and renames it by appending `.gz` to the filename
+     * @param string $filename the file to compress.
+     * @return string the name of the compressed file.
+     * @since 2.0.13
+     */
+    protected function compressFile($filename)
```

Comment:
Is it possible that multiple processes call `compressFile` for the same file concurrently?